sg_jira/hook.py

```diff
@@ -0,0 +1,393 @@
+            sg_note["content"],
+        )
+
+    def extract_jira_comment_data(self, jira_comment_body):
+        """Helper method to extract the FPTR note data from a Jira comment body."""
+
+        result = re.search(self.JIRA_COMMENT_REGEX, jira_comment_body, flags=re.S)
+
+        # if the Jira comment body doesn't match our regex, that means the comment could have been created from Jira
+        # so the data will be found from the comment itself rather than its body
+        if not result:
+            return None, jira_comment_body, None
+
+        author = result.group(2).strip()
+        # we need to make sure the author is associated with a current FPTR user
+        sg_user = self._shotgun.find_one(
+            "HumanUser",
+            [["name", "is", author]]
+        )
+        if not sg_user:
+            raise InvalidJiraValue(
+                "content",
+                jira_comment_body,
+                f"Invalid Jira Comment panel formatting. Unable to parse FPTR "
+                "author from '%s'" % author,
+            )
+
+        subject = result.group(1).strip()
+        # if we have any { or } in the title reject the value as it is likely
+        # to be an ill-formed panel block.
+        if re.search(r"[\{\}]", subject):
+            raise InvalidJiraValue(
+                "content",
+                jira_comment_body,
+                f"Invalid Jira Comment panel formatting. Unable to parse FPTR "
+                "subject from '%s'" % subject,
+            )
+        content = result.group(3).strip()
+
+        return subject, content, sg_user
+
+    def compose_jira_worklog_comment(self, sg_timelog):
+        """Helper method to compose the Jira Worklog comment from a FPTR timelog."""
+        return self.WORKLOG_BODY_TEMPLATE % (
+            sg_timelog["user"]["name"],
+            sg_timelog.get("description", ""),
+        )
+
+    def extract_jira_worklog_data(self, jira_worklog_comment):
+        """Helper method to extract the FPTR timelog data from a Jira Worklog comment."""
+
+        result = re.search(self.JIRA_WORKLOG_REGEX, jira_worklog_comment, flags=re.S)
+
+        # if the Jira comment body doesn't match our regex, that means the comment could have been created from Jira
+        # so the data will be found from the comment itself rather than its body
+        if not result:
+            return jira_worklog_comment, None
+
+        author = result.group(1).strip()
+        # we need to make sure the author is associated with a current FPTR user
+        sg_user = self._shotgun.find_one(
+            "HumanUser",
+            [["name", "is", author]]
+        )
+        if not sg_user:
+            raise InvalidJiraValue(
+                "content",
+                jira_worklog_comment,
+                f"Invalid Jira Worklog comment formatting. Unable to parse FPTR "
+                "author from '%s'" % author,
+            )
+
+        return result.group(2).strip(), sg_user
```

Comment:
Should we have an error handling around worklog comment regex or is it not worth it?

Reply:
I don't think we should have one
when you create a worklog from Jira and push it to FPTR, you won't have the user name in the worklog comment as it will be properly set in the "created" field
so the regex won't match but it's not an error, in that case you want the whole worklog comment to be pushed to FPTR